lib/libzfs/libzfs_diff.c

```diff
@@ -257,14 +258,34 @@ write_inuse_diffs_one(FILE *fp, differ_info_t *di, uint64_t dobj)
 	 */
 	fobjerr = get_stats_for_obj(di, di->fromsnap, dobj, fobjname,
 	    MAXPATHLEN, &fsb);
-	if (fobjerr && di->zerr != ENOENT && di->zerr != ENOTSUP)
+	if (fobjerr && di->zerr != ENOENT && di->zerr != ENOTSUP &&
+	    di->zerr != EINVAL)
 		return (-1);
 
+	if (fobjerr && (di->zerr == EINVAL || di->zerr == ENOENT)) {
```

Comment:
> like EINVAL, which can apparently come from a file with multiple hardlinks having the one whose name is referenced deleted.

It looks like this may be the only legitimate case.  The code which handles this on the kernel side is `zfs_obj_to_path_impl()->zfs_obj_to_pobj()` and it actually includes a [small comment](https://github.com/openzfs/zfs/blob/master/module/os/linux/zfs/zfs_znode.c#L2044-L2052) explaining this case.

Rather than always allowing `EINVAL` (which it looks like can be returned for a variety of reasons) why don't we update the check here to return `ENOLINK`.  Then we can handle this in a fashion similar to `ESTALE` by adding a case for `ENOLINK` to `get_stats_for_obj()` which sets a useful error message but returns 0.

nit: It's purely cosmetic but for consistency please stick to using `/* */` styled comments instead of `//`.

Reply:
I can take a stab at that, but fixing EINVAL while still dying on ENOENT (or vice-versa) solves nothing for me, so I'm going to push something handling both.

My bad, I got the impression that `/* */` were used for n>2 line comments into my head somewhere.

Reply:
Oh, I see, you were proposing modifying the kernel module to return different errors here.

Sure, I'll try that, but separate from that, I still feel it'd be reasonable to default to printing e.g. "I got back EINVAL for object 1234" and continue diffing, rather than aborting on the first instance, as that rather limits the usefulness of `zfs diff` - I'd bet money that these aren't the only cases that can arise in nonfatal circumstances, and I can't obviously see any way continuing to run `zfs diff` can result in mangling your data.

Reply:
You know, I tried just changing the `EINVAL` return case to `ENOLINK`, no other changes, and that has a bunch of ripple effects of causing other random errors to crop up (among them, `EOPNOTSUPP` creeps in from somewhere) - in addition, trying to detect only the cases of `ENOENT` that we want to change seems somewhat difficult (I spent a little bit looking at it, but the structures being passed around in `zfs_obj_to_pobj()` are somewhat opaque to me, so I didn't immediately see a good way to distinguish between causes).

Between those two things, I think I really prefer just handling it in libzfs_diff. So I've re-pushed that version, with the comments fixed.

Reply:
Yeah, I spent a little time looking in to this as well and it is a little trickier that I first thought.  I'm okay with this approach, but then there are other errors which should be handled.  For example, it looks like `EIO` is possible if the object is damaged and the checksum validation fails.  There are probably other cases as well. Since as you said it's never harmful to carry on, it seems like we'd want to print a warning for any of these unexpected errors.

I think it'd be reasonable to drop the other new comments and then try and summarize the high level policy by adding a little bit to the existing comment.  Something like this:

```
        /*
         * Check the from and to snapshots for info on the object. If
         * we get ENOENT, then the object just didn't exist in that
         * snapshot.  If we get ENOTSUP, then we tried to get info on
         * a non-ZPL object, which we don't care about anyway.  For any
         * other error we print a warning which includes the errno and
         * carry on.
         */
```

Reply:
Sorry about the delay, I kept getting caught up with other things; is something like what I pushed what you had in mind?

I ran it through two pairs of snapshots I have that spit out EINVAL and ENOENT - one on Linux/x86_64, one on FreeBSD/x86_64, and neither caused the earth to split open and release demons, though obviously that's not a complete enumeration of errno.

Reply:
Yes, that's what I mind.  You can probably use `zinject` to test the damaged dataset / EIO case.  Something like:

`log_must zinject -a -t data -e checksum -T read $FILEPATH`

Reply:
I ended up needing to use -t dnode -T all on the parent directory to get it to respond at all (with EBADE for -e checksum and EIO for -e io, which I think we'd expect here); all the permutations of -T read that I tried, on the parent directory or the file, on either or both snapshot, -t dnode or -t data or both, it didn't seem to notice.

Anyway, it finished printing the diff of my entire huge dataset without dying prematurely or doing anything unreasonable in the parts I saw; if it's reasonable to you and the bots don't tell me the moon is falling, I think I'm happy with it currently.